apps/condo/domains/miniapp/schema/preprocessors/B2BAppAccess.js

```diff
@@ -0,0 +1,470 @@
+        if (typeof access.create !== 'undefined') newAccess.create = fieldAccessWrapperIfNeeded(access.create, fnWrapper)
+        if (typeof access.update !== 'undefined') newAccess.update = fieldAccessWrapperIfNeeded(access.update, fnWrapper)
+        if (typeof access.delete !== 'undefined') newAccess.delete = fieldAccessWrapperIfNeeded(access.delete, fnWrapper)
+        if (typeof access.auth !== 'undefined') newAccess.auth = fieldAccessWrapperIfNeeded(access.auth, fnWrapper)
+        return newAccess
+    }
+
+    throw new Error(
+        `fieldAccessWrapperIfNeeded(), received ${type}.`,
+    )
+}
+
+/**
+ *
+ * @param pathToOrganizationId {string[]}
+ * @return {string}
+ *
+ * @example
+ * Input: ['a', 'b', 'c']
+ * Output: '{ a { b { c } } }'
+ */
+const generateGqlDataPart = (pathToOrganizationId) => {
+    if (!pathToOrganizationId || !isArray(pathToOrganizationId) || isEmpty(pathToOrganizationId)) throw new Error('"pathToOrganizationId" should not be empty array')
+    return `{ ${pathToOrganizationId.join(' { ')}${' }'.repeat(pathToOrganizationId.length)}`
+}
+
+const GENERATED_SCHEMAS = new Map()
+
+/**
+ *
+ * @param pathToOrganizationId {string[]}
+ * @return {*|DocumentNode}
+ *
+ * @example
+ * Input: ['ticket', 'organization', 'id']
+ * Output: gql`
+ *         query getAllTickets ($where: TicketWhereInput, $first: Int = 1) {
+ *             objs: allTickets(where: $where, first: $first) { organization { id } }
+ *         }
+ *     `
+ */
+const generateGql = (pathToOrganizationId) => {
+    if (!pathToOrganizationId || !isArray(pathToOrganizationId) || isEmpty(pathToOrganizationId)) throw new Error('"pathToOrganizationId" should not be empty array')
+    if (pathToOrganizationId.length < 3) throw new Error(`To generate gql "pathToOrganizationId" must contain at least three elements! But was ${pathToOrganizationId}`)
+
+    const gqlKey = pathToOrganizationId.join(':')
+
+    if (GENERATED_SCHEMAS.has(gqlKey)) {
+        return GENERATED_SCHEMAS.get(gqlKey)
+    }
+
+    const listKey = pathToOrganizationId[0].charAt(0).toUpperCase() + pathToOrganizationId[0].slice(1)
+
+    const generatedGql = gql`
+        query getAll${pluralize.plural(listKey)} ($where: ${listKey}WhereInput, $first: Int = 1) {
+            objs: all${pluralize.plural(listKey)}(where: $where, first: $first) ${generateGqlDataPart(pathToOrganizationId.slice(1))}
+        }
+    `
+    GENERATED_SCHEMAS.set(gqlKey, generatedGql)
+
+    return generatedGql
+}
+
+const canManageByServiceUser = async ({ authentication: { item: user }, listKey, originalInput, itemId, operation, context }, schemaConfig) => {
+    if (!user) return throwAuthenticationError()
+    if (user.deletedAt) return false
+
+    if (!listKey) return false
+
+    const pathToOrganizationId = get(schemaConfig, 'pathToOrganizationId', ['organization', 'id'])
+
+    let organizationId
+
+    if (operation === 'create') {
+        if (pathToOrganizationId.length === 1) {
```

Comment:
Why 1, 2, 3, error cases? Aren't there only 2 of them:

1. Direct access: Not sure that it should take id from original input (since it generated), but ok. Anyway this case is not real, since we prevent organization from being created.
2. Direct relation access: get it from connect.id is fine.
3. Non-direct relation access. (any length > 2). Get itemId from connect, do request returning organization id. Access it.

Reply:
I don't know of cases where we need to check for  connection `A -> B -> C -> D (-> ...)`. (And actually at the moment it is not needed) Therefore, I preferred to leave the implementation of these cases for later, if necessary

Reply:
refactored

Reply:
You've already wrote a helpful utils for this. So any depth can be done with an ease :)

Because right now it seems something like...
![image](https://github.com/open-condo-software/condo/assets/50069872/18ea4338-f70c-45b8-9e54-31ce6decffc3)